server/quota/quota_manager.go

```diff
@@ -0,0 +1,291 @@
+
+type QuotaManager struct {
+	env           environment.Env
+	namespaces    map[string]*namespace
+	bucketCreator bucketCreatorFn
+}
+
+func NewQuotaManager(env environment.Env) (*QuotaManager, error) {
+	return newQuotaManager(env, createGCRABucket)
+}
+
+func newQuotaManager(env environment.Env, bucketCreator bucketCreatorFn) (*QuotaManager, error) {
+	config, err := fetchConfigFromDB(env)
+	if err != nil {
+		return nil, err
+	}
+	qm := &QuotaManager{
+		env:           env,
+		namespaces:    make(map[string]*namespace),
+		bucketCreator: bucketCreator,
+	}
+
+	for nsName, nsConfig := range config {
+		ns, err := qm.createNamespace(env, nsName, nsConfig)
+		if err != nil {
+			return nil, err
+		}
+		qm.namespaces[nsName] = ns
+	}
+
+	return qm, nil
+}
+
+func (qm *QuotaManager) createNamespace(env environment.Env, name string, config *namespaceConfig) (*namespace, error) {
+	defaultBucketConfig := config.bucketsByName[defaultBucketName]
+	if defaultBucketConfig == nil {
+		return nil, status.InvalidArgumentErrorf("default quota bucket is unset in namespace: %q", name)
+	}
+
+	defaultBucket, err := qm.bucketCreator(env, defaultBucketConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	ns := &namespace{
+		name:          name,
+		config:        config,
+		defaultBucket: defaultBucket,
+		bucketsByKey:  make(map[string]Bucket),
+	}
+
+	for _, bc := range config.bucketsByName {
+		bucket, err := qm.bucketCreator(env, bc)
+		if err != nil {
+			return nil, err
+		}
+
+		keys := config.quotaKeysByBucketName[bc.Name]
+		for _, key := range keys {
+			ns.bucketsByKey[key] = bucket
+		}
+	}
+	return ns, nil
+}
+
+// findBucket finds the bucket given a namespace and key. If the key is found in
+// bucketsByKey map, return the corresponding bucket. Otherwise, return the
+// default bucket. Returns nil if the namespace is not found.
+func (qm *QuotaManager) findBucket(namespace string, key string) Bucket {
+	ns, ok := qm.namespaces[namespace]
+	if !ok {
+		log.Warningf("namespace %q not found", namespace)
+		return nil
+	}
+
+	if b, ok := ns.bucketsByKey[key]; ok {
+		return b
+	}
+
+	return ns.defaultBucket
+}
+
+func (qm *QuotaManager) getGroupID(ctx context.Context) string {
+	if a := qm.env.GetAuthenticator(); a != nil {
+		user, err := a.AuthenticatedUser(ctx)
+		if err != nil {
+			return interfaces.AuthAnonymousUser
+		}
+		return user.GetGroupID()
+	}
+	return ""
+}
+
+func (qm *QuotaManager) getKey(ctx context.Context) string {
+	groupID := qm.getGroupID(ctx)
+	if groupID != "" {
+		return groupID
+	}
+	//TODO(lulu): find IP address
+	return ""
+}
+
+func (qm *QuotaManager) Allow(ctx context.Context, namespace string, quantity int64) (bool, error) {
+	key := qm.getKey(ctx)
+	b := qm.findBucket(namespace, key)
+	if b == nil {
+		log.Warningf("quota bucket for namespace %q and key %q not found", namespace, key)
+		return true, nil
+	}
+	return b.Allow(ctx, key, quantity)
+}
+
+func Register(env environment.Env) error {
```

Comment:
do you want to put this behind a flag so we can flip it on / off as we roll it out?

Reply:
Done.